File: backend/src/test/java/org/globe42/web/charges/ChargeTypeControllerTest.java
package org.globe42.web.charges;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import java.math.BigDecimal;
import java.util.Collections;
import java.util.List;
import java.util.Optional;

import org.globe42.dao.ChargeCategoryDao;
import org.globe42.dao.ChargeTypeDao;
import org.globe42.domain.ChargeCategory;
import org.globe42.domain.ChargeType;
import org.globe42.test.BaseTest;
import org.globe42.web.exception.BadRequestException;
import org.globe42.web.exception.NotFoundException;
import org.junit.Before;
import org.junit.Test;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.InjectMocks;
import org.mockito.Mock;

/**
 * Unit tests for {@link ChargeTypeController}
 * @author JB Nizet
 */
public class ChargeTypeControllerTest extends BaseTest {
    @Mock
    private ChargeTypeDao mockChargeTypeDao;

    @Mock
    private ChargeCategoryDao mockChargeCategoryDao;

    @InjectMocks
    private ChargeTypeController controller;

    @Captor
    private ArgumentCaptor<ChargeType> chargeTypeArgumentCaptor;

    private ChargeType chargeType;

    @Before
    public void prepare() {
        chargeType = new ChargeType(42L);
        chargeType.setName("source 1");
        chargeType.setCategory(new ChargeCategory(1L, "category 1"));
        chargeType.setMaxMonthlyAmount(new BigDecimal("1234.56"));
    }

    @Test
    public void shouldList() {
        when(mockChargeTypeDao.findAll()).thenReturn(Collections.singletonList(chargeType));

        List<ChargeTypeDTO> result = controller.list();

        assertThat(result).hasSize(1);
        ChargeTypeDTO dto = result.get(0);
        assertThat(dto.getId()).isEqualTo(chargeType.getId());
        assertThat(dto.getName()).isEqualTo(chargeType.getName());
        assertThat(dto.getCategory().getId()).isEqualTo(chargeType.getCategory().getId());
        assertThat(dto.getCategory().getName()).isEqualTo(chargeType.getCategory().getName());
        assertThat(dto.getMaxMonthlyAmount()).isEqualTo(chargeType.getMaxMonthlyAmount());
        assertThat(dto.getId()).isEqualTo(chargeType.getId());
    }

    @Test
    public void shouldGet() {
        when(mockChargeTypeDao.findById(chargeType.getId())).thenReturn(Optional.of(chargeType));

        ChargeTypeDTO result = controller.get(chargeType.getId());

        assertThat(result.getId()).isEqualTo(chargeType.getId());
    }

    @Test(expected = NotFoundException.class)
    public void shouldThrowWhenGettingWithUnknownId() {
        when(mockChargeTypeDao.findById(chargeType.getId())).thenReturn(Optional.empty());

        controller.get(chargeType.getId());
    }

    @Test
    public void shouldCreate() {
        ChargeTypeCommandDTO command = createCommand();

        when(mockChargeCategoryDao.findById(command.getCategoryId()))
            .thenReturn(Optional.of(new ChargeCategory(command.getCategoryId(), "category 2")));
        when(mockChargeTypeDao.save(any(ChargeType.class))).thenReturn(chargeType);

        ChargeTypeDTO result = controller.create(command);

        verify(mockChargeTypeDao).save(chargeTypeArgumentCaptor.capture());

        assertThat(result).isNotNull();
        assertChargeTypeEqualsCommand(chargeTypeArgumentCaptor.getValue(), command);
    }

    @Test(expected = BadRequestException.class)
    public void shouldThrowWhenCreatingWithUnknownChargeCategory() {
        ChargeTypeCommandDTO command = createCommand();

        when(mockChargeCategoryDao.findById(command.getCategoryId()))
            .thenReturn(Optional.of(new ChargeCategory(command.getCategoryId(), "category 2")));
        when(mockChargeTypeDao.existsByName(command.getName())).thenReturn(true);

        controller.create(command);
    }

    @Test(expected = NotFoundException.class)
    public void shouldThrowWhenCreatingWithExistingName() {
        ChargeTypeCommandDTO command = createCommand();

        when(mockChargeCategoryDao.findById(command.getCategoryId())).thenReturn(Optional.empty());

        controller.create(command);
    }

    @Test
    public void shouldUpdate() {
        ChargeTypeCommandDTO command = createCommand();

        when(mockChargeTypeDao.findById(chargeType.getId())).thenReturn(Optional.of(chargeType));
        when(mockChargeCategoryDao.findById(command.getCategoryId()))
            .thenReturn(Optional.of(new ChargeCategory(command.getCategoryId(), "category 2")));

        controller.update(chargeType.getId(), command);

        assertChargeTypeEqualsCommand(chargeType, command);
    }

    @Test(expected = NotFoundException.class)
    public void shouldThrowIfNotFoundWhenUpdating() {
        when(mockChargeTypeDao.findById(chargeType.getId())).thenReturn(Optional.empty());

        controller.update(chargeType.getId(), createCommand());
    }

    @Test(expected = BadRequestException.class)
    public void shouldThrowWhenUpdatingWithAlreadyUsedNickName() {
        ChargeTypeCommandDTO command = createCommand();

        when(mockChargeTypeDao.findById(chargeType.getId())).thenReturn(Optional.of(chargeType));
        when(mockChargeCategoryDao.findById(command.getCategoryId()))
            .thenReturn(Optional.of(new ChargeCategory(command.getCategoryId(), "category 2")));
        when(mockChargeTypeDao.findByName(command.getName())).thenReturn(Optional.of(new ChargeType(4567L)));

        controller.update(chargeType.getId(), command);
    }

    @Test
    public void shouldUpdateIfSameSurnameIsKept() {
        ChargeTypeCommandDTO command = createCommand();

        when(mockChargeTypeDao.findById(chargeType.getId())).thenReturn(Optional.of(chargeType));
        when(mockChargeCategoryDao.findById(command.getCategoryId()))
            .thenReturn(Optional.of(new ChargeCategory(command.getCategoryId(), "category 2")));
        when(mockChargeTypeDao.findByName(command.getName())).thenReturn(Optional.of(chargeType));

        controller.update(chargeType.getId(), command);

        assertChargeTypeEqualsCommand(chargeType, command);
    }

    static ChargeTypeCommandDTO createCommand() {
        return new ChargeTypeCommandDTO("source 2", 2L, new BigDecimal("12.34"));
    }

    private void assertChargeTypeEqualsCommand(ChargeType source, ChargeTypeCommandDTO command) {
        assertThat(source.getName()).isEqualTo(command.getName());
        assertThat(source.getCategory().getId()).isEqualTo(command.getCategoryId());
        assertThat(source.getMaxMonthlyAmount()).isEqualTo(command.getMaxMonthlyAmount());
    }
}

File: backend/src/test/kotlin/org/globe42/web/charges/ChargeTypeControllerTest.kt
package org.globe42.web.charges

import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatExceptionOfType
import org.globe42.dao.ChargeCategoryDao
import org.globe42.dao.ChargeTypeDao
import org.globe42.domain.ChargeCategory
import org.globe42.domain.ChargeType
import org.globe42.web.exception.BadRequestException
import org.globe42.web.exception.NotFoundException
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.math.BigDecimal

/**
 * Unit tests for [ChargeTypeController]
 * @author JB Nizet
 */
class ChargeTypeControllerTest {
    private val mockChargeTypeDao = mockk<ChargeTypeDao>()
    private val mockChargeCategoryDao = mockk<ChargeCategoryDao>()
    private val controller = ChargeTypeController(mockChargeTypeDao, mockChargeCategoryDao)

    private lateinit var chargeType: ChargeType

    @BeforeEach
    fun prepare() {
        chargeType = ChargeType(42L).apply {
            name = "source 1"
            category = ChargeCategory(1L, "category 1")
            maxMonthlyAmount = BigDecimal("1234.56")
        }
    }

    @Test
    fun `should list`() {
        every { mockChargeTypeDao.findAll() } returns listOf(chargeType)

        val result = controller.list()

        assertThat(result).hasSize(1)
        val (id, name, category, maxMonthlyAmount) = result[0]
        assertThat(id).isEqualTo(chargeType.id)
        assertThat(name).isEqualTo(chargeType.name)
        assertThat(category.id).isEqualTo(chargeType.category.id)
        assertThat(category.name).isEqualTo(chargeType.category.name)
        assertThat(maxMonthlyAmount).isEqualTo(chargeType.maxMonthlyAmount)
        assertThat(id).isEqualTo(chargeType.id)
    }

    @Test
    fun `should get`() {
        val chargeTypeId = chargeType.id!!
        every { mockChargeTypeDao.findByIdOrNull(chargeTypeId) } returns chargeType

        val result = controller.get(chargeTypeId)

        assertThat(result.id).isEqualTo(chargeTypeId)
    }

    @Test
    fun `should throw when getting with unknown id`() {
        every { mockChargeTypeDao.findByIdOrNull(chargeType.id!!) } returns null

        assertThatExceptionOfType(NotFoundException::class.java).isThrownBy { controller.get(chargeType.id!!) }
    }

    @Test
    fun `should create`() {
        val command = createCommand()

        every { mockChargeTypeDao.existsByName(command.name) } returns false
        every { mockChargeCategoryDao.findByIdOrNull(command.categoryId) } returns
            ChargeCategory(command.categoryId, "category 2")
        every { mockChargeTypeDao.save(any<ChargeType>()) } returns chargeType

        val result = controller.create(command)

        verify {
            mockChargeTypeDao.save(withArg<ChargeType> {
                assertChargeTypeEqualsCommand(it, command)
            })
        }

        assertThat(result).isNotNull

    }

    @Test
    fun `should throw when creating with unknown charge category`() {
        val command = createCommand()

        every { mockChargeCategoryDao.findByIdOrNull(command.categoryId) } returns
            ChargeCategory(command.categoryId, "category 2")
        every { mockChargeTypeDao.existsByName(command.name) } returns true

        assertThatExceptionOfType(BadRequestException::class.java).isThrownBy { controller.create(command) }
    }

    @Test
    fun `should throw when creating with existing name`() {
        val command = createCommand()

        every { mockChargeTypeDao.existsByName(command.name) } returns true

        assertThatExceptionOfType(BadRequestException::class.java).isThrownBy { controller.create(command) }
    }

    @Test
    fun `should update`() {
        val command = createCommand()

        every { mockChargeTypeDao.findByName(command.name) } returns null
        every { mockChargeTypeDao.findByIdOrNull(chargeType.id!!) } returns chargeType
        every { mockChargeCategoryDao.findByIdOrNull(command.categoryId) } returns
            ChargeCategory(command.categoryId, "category 2")

        controller.update(chargeType.id!!, command)

        assertChargeTypeEqualsCommand(chargeType, command)
    }

    @Test
    fun `should throw if not found when updating`() {
        every { mockChargeTypeDao.findByIdOrNull(chargeType.id!!) } returns null

        assertThatExceptionOfType(NotFoundException::class.java).isThrownBy {
            controller.update(
                chargeType.id!!,
                createCommand()
            )
        }
    }

    @Test
    fun `should throw when updating with already used name`() {
        val command = createCommand()

        every { mockChargeTypeDao.findByIdOrNull(chargeType.id!!) } returns chargeType
        every { mockChargeCategoryDao.findByIdOrNull(command.categoryId) } returns
            ChargeCategory(command.categoryId, "category 2")
        every { mockChargeTypeDao.findByName(command.name) } returns ChargeType(4567L)

        assertThatExceptionOfType(BadRequestException::class.java).isThrownBy {
            controller.update(chargeType.id!!, command)
        }
    }

    @Test
    fun `should update if same name is kept`() {
        val command = createCommand()

        every { mockChargeTypeDao.findByIdOrNull(chargeType.id!!) } returns chargeType
        every { mockChargeCategoryDao.findByIdOrNull(command.categoryId) } returns
            ChargeCategory(command.categoryId, "category 2")
        every { mockChargeTypeDao.findByName(command.name) } returns chargeType

        controller.update(chargeType.id!!, command)

        assertChargeTypeEqualsCommand(chargeType, command)
    }

    private fun assertChargeTypeEqualsCommand(source: ChargeType, command: ChargeTypeCommandDTO) {
        assertThat(source.name).isEqualTo(command.name)
        assertThat(source.category.id!!).isEqualTo(command.categoryId)
        assertThat(source.maxMonthlyAmount).isEqualTo(command.maxMonthlyAmount)
    }
}

internal fun createCommand(): ChargeTypeCommandDTO {
    return ChargeTypeCommandDTO("source 2", 2L, BigDecimal("12.34"))
}
